docs/design/sql/02-operator-interface.md

```diff
@@ -0,0 +1,157 @@
+    Row next();   // Get the next row
+    void close(); // Close the operator and release all resources
+}
+```
+
+## Volcano Model in Hazelcast Mustang
+
+The original Volcano Model has two drawbacks:
+1. Operators exchange one row at a time, which leads to performance overhead
+2. Call to the `next()` is blocking, which is not optimal for the distributed environment, where
+operators often wait for remote data or free space in the send buffer.
+
+To achieve high performance, we introduce several changes to the original Volcano Model: batching and
+non-blocking execution.
+
+### Row and RowBatch
+We define the `RowBatch` interface which a collection of rows (tuples).
+
+*Snippet 5: RowBatch interface*
+```java
+interface RowBatch {
+    Row getRow(int index); // Get the row by index
+    int getRowCount();     // Get the number of rows 
+} 
+```
+
+Then we define the `Row` interface, which provides access to values by index. The `Row` itself is considered
+as a special case of `RowBatch` with one row. This allows saving on allocations in some parts of the engine.
+
+*Snippet 6: Row interface*
+```java
+interface Row extends RowBatch {
+    Object get(int index); // Get the value by index
+    int getColumnCount();  // Get the number of values in the row 
+    
+    default int getRowCount() {
+        return 1;
+    }
+    
+    default int getRow(int index) {
+        return this;
+    }
+}
+```
+
+### Operator
+The operator is defined by `Exec` interface:
+1. Operators exchange `RowBatch` instead of `Row`
+1. The blocking `next()` method is replaced with the non-blocking `advance()` method, which returns the iteration
+result instead of the row batch
+1. The `RowBatch` could be accessed through a separate method
+1. The `open()` method is renamed to `setup()`. Special query context is passed to it as an argument
+1. There is no separate `close()` method because the engine doesn't need explicit per-operator cleanup at the
+moment. This may change in future, in this case the current document should be updated accordingly
+
+*Snippet 7: Exec interface*
+```java
+interface Exec {
+    void setup(QueryContext context); // Initialize the operator
+    IterationResult advance();        // Advance the operator if possible; never blocks
+    RowBatch currentBatch();          // Get the batch returned by the previous advance() call 
+}
+```
+
+The result of iteration is defined in the `IterationResult` enumeration.
+
+*Snippet 8: IterationResult enumeration*
+```java
+enum IterationResult {
+    FETCHED,      // Iteration produced new rows
+    FETCHED_DONE, // Iteration produced new rows and reached the end of the stream, no more rows are expected
+    WAIT;         // Failed to produce new rows, release the control
+}
+```
+
+When the engine has received `FETCHED` or `FETCHED_DONE` from the `Exec.advance()` call, it may access the
+produced rows through the `Exec.currentBatch()` call. If the engine has received `WAIT`, then query
```

Comment:
It's not clear who is the owner the returned row batch. May the consumer assume the row batch is immutable and the the ownership is transferred to the consumer? Or should the consumer assume that the row batch content may change at any/some time and the ownership is still held by the producer?

Reply:
This is a difficult question. I think that in the current prototype we are not concerned with it. Both approaches, with and without ownership transfer has their pros and cons:
1) With ownership transfer - easier to implement, but produce more litter, because every operator must produce its own copy of the batch
2) Without ownership transfer - hard to reason about memory (especially if we move to offheap eventually), but more performant

Another interesting point is that this question will not be critical as soon as we have a compilation because memory management decisions could be made at the fine-grained level there.

Also if we switch to some byte[] representations of rows, then copying would be required anyway.

That said I am thinking that implementation with ownership transfer is slightly better for us. What do you think?

Reply:
To my understanding, it's better to state explicitly that the ownership is still held by the producer and the row batch content is valid until the next `advance()` call. The consumer can always make a copy, if it needs one; the producer may be implemented in a "zero-allocation" way.

Reply:
... hmm, and the same question applies to `Row`

Reply:
I meant the same but shuffled the terms. Please read "with" as "without" and vice versa :-)
The row is assumed to be immutable.

Reply:
I updated the original comment.

Reply:
Clarifications about the life cycle have been added to the document. See "Implementation Guidelines".

Reply:
Added more clarifications (https://github.com/hazelcast/hazelcast/pull/16763/commits/f91e6967d4faa6c3ff5d25e121d60236bf1acf8a)

Reply:
It might make sense to work with rows in the same way. For instance, most aggregation functions don't require the rows themselves, so the producer might reuse the same preallocated set of rows other and other again. But it's not clear how to define the ownership transfer/bounds for rows.

Reply:
Aggregation is a rare exception where the operator cannot own the rows. Most other operators require full control of the row state - filter, project, join, sort, etc..

Reply:
Probably "the same way" was the wrong wording. I was thinking about the situation when we, for instance, doing a simple scan producing 1M rows, we end up allocating 1M rows in total. But in reality we send the rows in small batches, batch by batch. So we really need only batch size rows allocated, once the batch rows hit the wire or consumed in some other way, we can reuse the same set of rows.

Reply:
It's impossible in the general case. 

First, we need to track row usage and sending somehow, which is not the case at the moment and would require additional code infrastructure.

Second, this is merely an implementation detail that we use `HeapRow` by default, pretty unfortunate detail. In future we likely to switch to `byte[]`-based or `byte*`-based implementation, where no row reuse is possible at all in the general case. Instead he reusable resource will be the byte array (or at least row batch), but not row.

Last, batch allocations depend on the operator. `Project` operator may split the result set into batches and align it with the sender batch size. But blocking operators, such as `Sort`, cannot do that.